pennylane/tape/tapes/tape.py

```diff
@@ -281,6 +286,7 @@ def __exit__(self, exception_type, exception_value, traceback):
             self._stack.__exit__(exception_type, exception_value, traceback)
 
         self._process_queue()
+        QuantumTape._lock.release()
```

Comment:
My only suggestion here: maybe it's better to use the context manager approach, or a `try`/`finally` block, to make sure that the thread is released if something goes wrong?

For example,

```python
with QuantumTape._lock:
    # queuing logic
```

or 
```python
QuantumTape._lock.acquire()
try:
    # queuing logic
finally:
    QuantumTape._lock.release()
```

Reply:
Thank you. The point here is to unlock after `with` block completion. Therefore locking/unlocking has to be in the different functions:  `__enter__` and `__exit__`. As you mentioned, I have to consider the case where queueing logic fails because of some failure. In that case, the error already corrupts the queueing operation, so as you suggested releasing the lock and throw the same error caught will be the option. I will fix the issue.

Reply:
Oh, my mistake @shukob, I forgot that the thread locking and releasing is already taking place inside `__enter__` and `__exit__`; so your current implementation should be okay

Reply:
No problem! I found that exception of queueing may cause a deadlock if that exception is handled in another place in that thread. Thank you for pointing me that.